Add i18n page (#17)
forem/forem-docs: docs/technical-overview/i18n.md

---
sidebar_position: 7
---

# Internationalization (i18n)

Forem's internationalization is currently in its early stages. You are *encouraged* to follow patterns laid out here, but it is not yet *required* for every pull request until further development and stabilization of the feature.

> This development is a matter of translation as well as de-coupling of copy from code. As this functionality is new, best practices in this area are still a work in progress.

Reviewer: _One more quick suggestion!:_

```suggestion
This development is a matter of translation as well as decoupling of copy from code. As this functionality is new, best practices in this area are still a work in progress.
```

> ## I18n features and direction

Reviewer: _Quick, non-blocking question:_ should the "i" in "i18n" here be lowercase?
```suggestion
## i18n features and direction
```

## Contributing

The most helpful contributions are anything that gets the project a step closer to full localization. A PR that tries to do it all at once will not be as helpful. Any opportunity to _remove_ copy entirely that is not necessary is also a step in the right direction and can be considered at pull-request time as an option, though that has a higher chance of being rejected due to the change to product functionality vs an objective improvement to localization.

We use [standard Rails localization practices](https://guides.rubyonrails.org/i18n.html). We have files related to each locale such as `config/locales/fr.yml`.

The yml file will look like this...

```yml
---
fr:
  core:
    add_comment: Ajouter un commentaire
    comment: Commentaire
    copy_link: Copier le lien
    create_account: Créer un profil
```

And in HTML, these can be used as helpers, such as...

```html
<div><%= t("core.add_comment") %></div>
```

`t` in this view helper context is short for `I18n.translate`. Check out the linked Rails guide for full information about how to use all standard translation and localization functionality.

### i18n.js

> We use the [i18n.js](https://github.com/fnando/i18n-js) gem to pass values to JavaScript. Our frontend code is currently divided into old asset pipeline code in `app/assets/javascripts` and modern Rails JS practices at `app/javascript`. In the asset pipeline you should directly use `I18n.t` and in app/javascript, there is a wrapper at `app/javascript/utilities/locale.js` you can include as you go. This is all subject to change in the name of code quality improvements.

Reviewer: _**Quick, non-blocking suggestion:**_

```suggestion
We use the [i18n.js](https://github.com/fnando/i18n-js) gem to pass values to JavaScript. Our frontend code is currently divided into old asset pipeline code in `app/assets/javascripts` and modern Rails JS practices at `app/javascript`. In the asset pipeline you should directly use `I18n.t` and in `app/javascript`, there is a wrapper at `app/javascript/utilities/locale.js` you can include as you go. This is all subject to change in the name of code quality improvements.
```

> This should generally be abstracted away for you if you follow above patterns, but for your information, we store translatiions in the view like this:

Reviewer: _Quick, typo fix:_

```suggestion
This should generally be abstracted away for you if you follow above patterns, but for your information, we store translations in the view like this:
```
